### PR TITLE
Enable timeout for DDPStrategy

### PR DESCRIPTION
**Pull request by lsy643**
## What does this PR do?

Enable timeout for DDPStrategy

Fixes https://github.com/PyTorchLightning/pytorch-lightning/issues/13211

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/strategies/ddp.py [hunk 1]
```diff
@@ -87,6 +89,7 @@ def __init__(
         ddp_comm_wrapper: Optional[callable] = None,
         model_averaging_period: Optional[int] = None,
         process_group_backend: Optional[str] = None,
+        timeout: Optional[timedelta] = default_pg_timeout,
```
Comment (carmocca) [inline pytorch_lightning/strategies/ddp.py:92]: Can you update the CHANGELOG? In the "Added" section

**Comment by justusschock** [comment]
@carmocca should we also set the env variables for the non-gloo backends? Or at least check them and error out early?

File: pytorch_lightning/strategies/ddp.py (hunk 1 above)
Comment (lsy643) [reply pytorch_lightning/strategies/ddp.py:92]: @carmocca OK.

File: pytorch_lightning/strategies/ddp.py [hunk 2]
```diff
@@ -202,7 +206,7 @@ def setup_distributed(self):
         rank_zero_only.rank = self.global_rank
 
         self._process_group_backend = self._get_process_group_backend()
-        init_dist_connection(self.cluster_environment, self._process_group_backend)
+        init_dist_connection(self.cluster_environment, self._process_group_backend, timeout=self._timeout)
```
[the remark this answers is not in the document]
Comment (lsy643) [reply pytorch_lightning/strategies/ddp.py:209]: @carmocca OK. A test has been added

**Review by justusschock — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]

**Comment by RuABraun** [comment]
How is one supposed to use this in a yaml config? the timeout arg is class with a named arg.

**Pull request merged**